yarn/src/main/scala/org/apache/spark/deploy/yarn/YarnAllocator.scala

```diff
@@ -81,8 +82,20 @@ private[yarn] class YarnAllocator(
     new ConcurrentHashMap[ContainerId, java.lang.Boolean])
 
   @volatile private var numExecutorsRunning = 0
-  // Used to generate a unique ID per executor
-  private var executorIdCounter = 0
+
+  /**
+   * Used to generate a unique ID per executor
+   *
+   * Init `executorIdCounter`. when AM restart, `executorIdCounter` will reset to 0. Then
+   * the id of new executor will start from 1, this will conflict with the executor has
+   * already created before. So, we should initialize the `executorIdCounter` by getting
+   * the max executorId from driver.
+   *
+   * @see SPARK-12864
+   */
+  private var executorIdCounter: Int = {
+    driverRef.askWithRetry[Int](RetrieveMaxExecutorId) + 1
```

Comment:
Hi @jerryshao , I made a test just now. It can work in yarn cluster mode. From my understanding, `driverRef` is created by `ApplicationMaster`, so, I think at this time `driverEndpoint` is started. What's your opinion ? Thanks a lot.